language/tools/move-package/src/resolution/resolution_graph.rs

```diff
@@ -0,0 +1,599 @@
+        dep_name: &PackageName,
+        mut root_path: PathBuf,
+    ) -> Result<(SourceManifest, PathBuf)> {
+        root_path.push(&dep.local);
+        match std::fs::read_to_string(&root_path.join(SourcePackageLayout::Manifest.path())) {
+            Ok(contents) => {
+                let source_package: SourceManifest =
+                    parse_move_manifest_string(contents).and_then(parse_source_manifest)?;
+                Ok((source_package, root_path))
+            }
+            Err(_) => Err(anyhow::format_err!(
+                "Unable to find package manifest for '{}' at {:?}",
+                dep_name,
+                SourcePackageLayout::Manifest.path().join(root_path),
+            )),
+        }
+    }
+}
+
+impl ResolvingPackage {
+    // Extend and check for duplicate names in rename_to
+    fn extend_renaming(
+        renaming: &mut Renaming,
+        dep_name: &PackageName,
+        dep_renaming: Renaming,
+    ) -> Result<()> {
+        for (rename_to, rename_from) in dep_renaming.into_iter() {
+            // We cannot rename multiple named addresses to the same name. In the future we'll want
+            // to support this.
+            if renaming.insert(rename_to, rename_from).is_some() {
+                bail!(
+                    "Duplicate renaming of named address '{}' found in dependency '{}'",
+                    rename_to,
+                    dep_name
+                );
+            }
+        }
+        Ok(())
+    }
+
+    // The resolution table contains the transitive closure of addresses that are known in that
+    // package. Extends the package's resolution table and checks for duplicate renamings that
+    // conflict during this process.
+    fn extend_resolution_table(
+        resolution_table: &mut ResolvingTable,
+        dep_name: &PackageName,
+        dep_resolution_table: ResolvingTable,
+        dep_renaming: Renaming,
+    ) -> Result<()> {
+        let renames = dep_renaming
+            .into_iter()
+            .map(|(rename_to, (_, rename_from))| (rename_from, rename_to))
+            .collect::<BTreeMap<_, _>>();
+
+        for (addr_name, addr_value) in dep_resolution_table.into_iter() {
+            let addr_name = renames.get(&addr_name).cloned().unwrap_or(addr_name);
+            if let Some(other) = resolution_table.insert(addr_name, addr_value.clone()) {
+                // They need to be the same refcell so resolve to the same location if there are any
+                // possible reassignments
+                if other.value != addr_value.value {
+                    bail!(
+                        "Named address '{}' in dependency '{}' is already set to '{}' but was then reassigned to '{}'",
+                        &addr_name,
+                        dep_name,
+                        match other.value.take() {
+                            None => "unassigned".to_string(),
+                            Some(addr) => format!("0x{}", addr.short_str_lossless()),
```

Comment:
Not a big deal for now, but keep in mind that in the source language, address values can be decimal too. We should maybe honor/extend that to other tools if possible. But really a long term improvement, not worth fussing with now